src/storage/nats_kv.rs

```diff
@@ -0,0 +1,304 @@
+//! Storage engine backed by NATS Kv
+//!
+//! This storage engine enables storing any WADM related information using NATS Kv as a backend
+
+use async_trait::async_trait;
+use std::path::PathBuf;
+
+use async_nats::{
+    jetstream::kv::Config as JetstreamKvConfig, jetstream::kv::Store as JetstreamKvStore,
+    Client as NatsClient, ConnectOptions as NatsConnectOptions,
+};
+
+use tokio::fs::read as read_file;
+
+use derive_builder::Builder;
+use semver::{BuildMetadata, Prerelease, Version};
+use serde::ser::StdError;
+use serde::{Deserialize, Serialize};
+use serde_json::{from_slice, to_vec, Error as SerdeJsonError};
+
+use super::{EngineMetadata, LatticeStorage, StorageEngineError, StorageError, StoreOptions};
+
+use super::state::{LatticeState, LatticeStateBuilder, LatticeStateBuilderError};
+
+use crate::storage::{Store, StoreRead, StoreReadWrite, StoreWrite, WithStateMetadata};
+
+const STORAGE_ENGINE_NAME: &str = "nats_kv";
+const STORAGE_ENGINE_VERSION: Version = Version {
+    major: 0,
+    minor: 1,
+    patch: 0,
+    pre: Prerelease::EMPTY,
+    build: BuildMetadata::EMPTY,
+};
+
+////////////
+// Engine //
+////////////
+
+#[derive(Debug)]
+pub struct NatsKvStorageEngine {
+    /// Configuration for the NATS KV storage engine
+    config: NatsKvStorageConfig,
+
+    /// NATS client
+    client: NatsClient,
+}
+
+#[derive(Debug, Serialize, Deserialize, Builder)]
+pub struct NatsKvStorageConfig {
+    /// URL to the nats instance
+    pub nats_url: String,
+
+    /// Prefix to use with lattice buckets
+    pub lattice_bucket_prefix: Option<String>,
+
+    /// Authentication config for use with NATS
+    pub auth: Option<NatsAuthConfig>,
+}
+
+#[derive(Debug, Serialize, Deserialize, Clone)]
+pub struct NatsAuthConfig {
+    pub creds_file: Option<String>,
+    pub jwt_seed: Option<String>,
+    pub jwt_path: Option<PathBuf>,
+}
+
+impl From<nkeys::error::Error> for StorageEngineError {
+    fn from(err: nkeys::error::Error) -> StorageEngineError {
+        StorageEngineError::Engine(format!("failed to process nkeys: {err}"))
+    }
+}
+
+impl From<std::io::Error> for StorageEngineError {
+    fn from(err: std::io::Error) -> StorageEngineError {
+        StorageEngineError::Engine(format!("unexpected I/O error: {err}"))
+    }
+}
+
+async fn build_nats_options(
+    cfg: &NatsAuthConfig,
+) -> Result<NatsConnectOptions, StorageEngineError> {
+    // Use JWT + seed if present
```

Comment:
Super nit: Might be easier to rewrite this function to have a single match block with each arm returning the connect options